File: lib/formats/gfx/xyz_converter.rb
require 'lib/binary_io'
require 'lib/image'
require 'zlib'

# Converts XYZ to PNG.
# Engine: RPGMaker
# Known games:
# - Yume Nikki
module XyzConverter
  module_function

  MAGIC = 'XYZ1'

  def add_cli_help(_arg_parser) end

  def parse_cli_options(_arg_parser, _options) end

  def decode!(file, _options)
    Decoder.new.read(file.data).update_file(file)
  end

  def encode!(_file, _options)
    fail 'Not supported.'
  end

  class Decoder
    def read(data)
      input = BinaryIO.from_string(data)

      magic = input.read(MAGIC.length)
      fail RecognitionError, 'Not a XYZ picture' if magic != MAGIC

      width, height = input.read(4).unpack('S2')

      input = BinaryIO.from_string(Zlib.inflate(input.read))
      palette = input.read(256 * 3).chunks(3)

      pixels = BinaryIO.from_string('')
      pixels.write(palette[input.read(1).ord]) until input.eof?
      pixels.rewind
      pixels = pixels.read

      Image.from_pixels(width, height, pixels, 'RGB')
    end
  end
end

File: src/formats/gfx/xyz_converter.cc
// XYZ image
//
// Company:   Enterbrain
// Engine:    RPGMaker
// Extension: .xyz
// Archives:  -
//
// Known games:
// - Yume Nikki

#include <cassert>
#include <stdexcept>
#include "formats/gfx/xyz_converter.h"
#include "formats/image.h"
#include "io.h"
#include "string_ex.h"

namespace
{
    const std::string magic("XYZ1", 4);
}

void XyzConverter::decode_internal(VirtualFile &file) const
{
    if (file.io.read(magic.size()) != magic)
        throw std::runtime_error("Not an XYZ image");

    uint16_t width = file.io.read_u16_le();
    uint16_t height = file.io.read_u16_le();

    size_t compressed_data_size = file.io.size() - file.io.tell();
    std::string compressed_data = file.io.read(compressed_data_size);
    std::string uncompressed_data = zlib_inflate(compressed_data);
    assert(uncompressed_data.size() == (unsigned)256 * 3 + width * height);

    size_t pixels_size = width * height * 3;
    std::unique_ptr<char> pixels(new char[pixels_size]);

    {
        const char *palette = uncompressed_data.data();
        const char *palette_indices = palette + 256 * 3;
        char *out = pixels.get();

        int i;
        size_t j;
        for (i = 0; i < width * height; i ++)
        {
            j = *palette_indices ++;
            *out ++ = palette[j * 3 + 0];
            *out ++ = palette[j * 3 + 1];
            *out ++ = palette[j * 3 + 2];
        }
    }

    std::unique_ptr<Image> image = Image::from_pixels(
        width,
        height,
        std::string(pixels.get(), pixels_size),
        IMAGE_PIXEL_FORMAT_RGB);
    image->update_file(file);
}
